src/core/net/ip6.cpp

```diff
@@ -1492,6 +1502,66 @@ bool Ip6::IsOnLink(const Address &aAddress) const
     return rval;
 }
 
+void Ip6::UpdateBorderRoutingCounters(const Message &  aMessage,
+                                      const Header &   aHeader,
+                                      const Header *   aInnerHeader,
+                                      bool             aForwardHost,
+                                      bool             aForwardThread,
+                                      const OptionMpl &aMplOption)
+{
+    otPacketsAndBytes *counter = nullptr;
+
+    if (!Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetSource()) &&
+        !Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetDestination()) &&
+        !aHeader.GetSource().IsLinkLocal() && !aHeader.GetDestination().IsLinkLocal() &&
+        aHeader.GetSource().GetPrefix() != Get<Mle::Mle>().GetMeshLocalPrefix() &&
+        aHeader.GetDestination().GetPrefix() != Get<Mle::Mle>().GetMeshLocalPrefix())
+    {
+        if (aForwardThread)
+        {
+            if (aHeader.GetDestination().IsMulticast())
+            {
+                // Do nothing. Inbound Multicast is handled later.
+            }
+            else
+            {
+                counter = &mBorderRoutingCounters.mInboundUnicast;
+            }
+        }
+        if (aForwardHost)
+        {
+            if (aHeader.GetDestination().IsMulticast())
+            {
+                if (aHeader.GetDestination().IsMulticastLargerThanRealmLocal())
+                {
+                    counter = &mBorderRoutingCounters.mOutboundMulticast;
+                }
+            }
+            else
+            {
+                counter = &mBorderRoutingCounters.mOutboundUnicast;
+            }
+        }
+    }
+
+    // Inbound Multicast
+    if (aForwardThread && aHeader.GetDestination().IsRealmLocalAllMplForwarders() && aInnerHeader &&
+        aInnerHeader->GetDestination().IsMulticastLargerThanRealmLocal() &&
+        aMplOption.GetSeedId() == Get<Mle::Mle>().GetRloc16() &&
```

Comment:
I need to check seed ID to verify if it's this device which initiates the MPL Data message. Therefore I had to extract the MPL option which was processed in `HandleExtensionHeaders` and I added an out parameter for `HandleExtensionHeaders` and `HandleOptions`. Is there a better way to achieve this?